src/python/createrepo_c/__init__.py

```diff
@@ -358,6 +362,273 @@ def decompress_file(src, dst, comtype, stat=None):
 detect_compression  = _createrepo_c.detect_compression
 compression_type    = _createrepo_c.compression_type
 
+# both the path and the *XmlFile objects need to be tracked together because there's no way to get the path
+# back from the *XmlFile objects
+MetadataInfoHolder = namedtuple("MetadataInfoHolder", ["path", "writer"])
+
+class RepositoryWriter:
+
+    _FINISHED_ERR_MSG = "Cannot perform action after the repository has already finished being written"
+
+    def __init__(self,
+                 destination,
+                 num_packages=None,
+                 unique_md_filenames=True,
+                 with_sqlite=False,
+                 changelog_limit=None,
+                 preserve_existing_metadata=False,
+                 metadata_compression=GZ_COMPRESSION,
+                 general_compression=GZ_COMPRESSION,
+                 metadata_checksum_type=SHA256,
+                 package_checksum_type=SHA256
+                ):
+        # TODO: with_zchunk option
+
+        if changelog_limit:
+            assert isinstance(changelog_limit, int) and changelog_limit >= 0, "changelog_limit must be an integer >= 0"
+
+        self._destination_repo_path = Path(destination)
+        self._repomd = Repomd()
+
+        self._unique_md_filenames = unique_md_filenames
+        self._with_sqlite = with_sqlite
+        self._changelog_limit = changelog_limit
+        self._preserve_existing_metadata = preserve_existing_metadata
+        self._metadata_checksum_type=metadata_checksum_type
+        self._package_checksum_type=package_checksum_type
+
+        self._has_set_num_pkgs = False
+        self._finished = False
+
+        # self._working_dir = tempfile.NamedTemporaryDirectory(prefix=".repodata", dir=destination)
+
+        # if self.destination.exists() and self.destination.is_dir():
+        #     if not os.path.isdir(DESTDIR):
+        #         raise IOError("%s is not a directory" % DESTDIR)
+        #     shutil.rmtree(DESTDIR)
+        # os.makedirs(destination, exist_ok=True)
+
+        os.makedirs(self.path, exist_ok=True)
+
+        # TODO: move this to finish() so that the existing repodata isn't moved until the new metadata has been
+        # successfully written?
+        # TODO: update option - copy over existing tags, packages to the new metadata
+        if self.repodata_path.exists():
+            if preserve_existing_metadata:
+                x = 0
+                while True:
+                    new_repodata_path = f"{self.repodata_path}_{x}"
```

Comment:
I think `createrepo_c` binary and `RepositoryWriter` should be as consistent as possible. How about using something closer to `--retain-old-md` and `--keep-all-metadata` instead of `preserve_existing_metadata`? 

Or do you prefer this behavior?

Reply:
This is just what the example had been doing when I copied it over.  I think there's a good argument to be made either way.  It does seem a little silly to keep the metadata files around but throw away the repomd.xml that tied them together.  If you keep the old repodata directories, it's fairly straightforwards to turn it back into a functional repo (assuming the packages are still there)